docs/configuration.md
# Configuration

CLI options and environment vars are equal in their usage. If both a CLI flag and an environment variable is set, the CLI flag takes precedence. For boolean CLI flags, they have no value. For environment vars, set them to either `true` or `false`, other values will not be accepted, and omitting them will set them to `false`.

## Required

- `--key` / `CAMO_KEY` - Randomly generated string used as a key for calculating the HMAC digest.
- `--root-url` / `CAMO_ROOT_URL` - URL, including a trailing slash, relative to the domain Camo is running on. For example, if Camo is available on `example.com/camo/`, set this to `/camo/`. For installations that do not run in a subdirectory, set this to `/`.

## Allowed content-types

At least one `content-type` needs to be allowed, or Camo will refuse to start.

- `--allow-audio` / `CAMO_ALLOW_AUDIO` - Whether `audio/*` MIME types should be allowed. (default: `false`)
- `--allow-image` / `CAMO_ALLOW_IMAGE` - Whether `image/*` MIME types should be allowed. (default: `false`)
- `--allow-video` / `CAMO_ALLOW_VIDEO` - Whether `video/*` MIME types should be allowed. (default: `false`)

Alternatively, you can set `--allow-all-types` / `CAMO_ALLOW_ALL_TYPES` (default: `false`), which will completely bypass any `content-type` checks, and thus allows all responses, even ones with a missing `content-type`.

## Other settings

- `--header-via` / `CAMO_HEADER_VIA` - The string used to identify this `camo-rs` instance in upstream requests. (default: `camo-rs asset proxy (+https://github.com/denschub/camo-rs)`)
- `--length-limit` / `CAMO_LENGTH_LIMIT` - The maximum `content-length` proxied by `camo-rs`. (default: `52428800` (50 MiB))
- `--listen` / `CAMO_LISTEN` - IP and Port this application should listen on. (default: `[::]:8081`)
- `--threads` / `CAMO_THREADS` - The number of worker threads to use. (default: the number of available CPU cores)
- `--upstream-timeout` / `CAMO_UPSTREAM_TIMEOUT` - The number of seconds to wait for an upstream response. (default: `10`)

## Logging

By default, `camo-rs` is very quiet. It will only ever say anything if something goes wrong. Optional logging is available.

### Log levels

The `--log-level` flag or `CAMO_LOG_LEVEL` env var can have the following values:

- `quiet` - Doesn't log anything at all, unless something unexpected goes wrong. This is the default.
- `warn` - Logs when valid requests couldn't be processed due to upstream errors, or if requests have been blocked by length limits or content-type restrictions.
- `info` - Logs the same was `warn`, but additionally logs when the request field encoding was wrong, or if the HMAC was invalid.

### Log formats

The `--log-format` flag or `CAMO_LOG_FORMAT` env var can be set to:

- `text` - Does log the information in a human-readable format.
- `text-color` - The same as `text`, but with some color added to improve readability. This is the default.
- `json` - Logs the data in a machine-readable JSON format, with one log entry per line.


Changelog.md
# 1.4.0

This versions introduces a new setting, `--threads`/`CAMO_THREADS` that allows limiting the number of worker threads spawned by `camo-rs`. If this flag is not set, the number of available CPU cores will be used, which matches the current behavior.

# 1.3.5

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.3.4

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.3.3

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.3.2

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.3.1

This version does not contain any functional changes. It updates third-party dependencies and enables more compile-time optimizations. For most users, the only difference should be more consistent binary sizes, but there might be minimal performance improvements in larger deployments.

# 1.3.0

This version includes the following non-breaking changes:

- It adds a response to `/robots.txt`, disallowing all robots from interacting with `camo-rs` in any way. This was inspired by a recent uptick of bots that crawl Camo instances. Not all bots to listen to those instructions, but some do.
- It re-ships `jemalloc` inside the Docker container, but doesn't enable it. If you run the Docker container and want to use `jemalloc`, you can add a `LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2` to the environment variables, and it will automatically load it. Likewise, tuning `jemalloc` with the `MALLOC_CONF` environment variable is posisble,too.

# 1.2.2

This version does not contain any functional changes. It only updates third-party dependencies. Included are two updates that address security issues:

- an update to `h2` to resolve [a "moderate severity" Denial of Service attack](https://github.com/advisories/GHSA-8r5v-vm4m-4g25)
- an update to `mio` to resolve [a "high severity" Windows-only vulnerability](https://github.com/advisories/GHSA-r8w9-5wcg-vfj7)

# 1.2.1

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.2.0

This version adds a new setting, `--allow-all-types` / `CAMO_ALLOW_ALL_TYPES` (default: `false`). This allows you to use `camo-rs` with all upstream `content-type`s, and even passes responses that do not have a `content-type` set.

# 1.1.8

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.1.7

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.1.6

This version does not contain any functional changes.

# 1.1.5

This version does not contain any functional changes. It only updates third-party dependencies. Included in those is an update for `h2` to resolve [a potential Denial of Service attack](https://nvd.nist.gov/vuln/detail/CVE-2023-26964) when dealing with http/2 upstream connections.

# 1.1.4

- Offical Docker image: Drop `libjemalloc`. There have been two reports about increased idle memory usage when using `libjemalloc`, having `camo-rs` sit above the 100 MiB mark. While the allocator would release that memory under memory pressure, this application isn't designed for a usage where allocator performance is important anyway, so let's move back to the default system allocator for the official container image.

# 1.1.3

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.1.2

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.1.1

This version does not contain any functional changes. It only updates third-party dependencies.

# 1.1.0

- Switch from OpenSSL-bindings to Rustls, to avoid incompatibilities between OpenSSL 1.x and 3.x (looking at you, Canonical)...

# 1.0.0

This is the first release version of `camo-rs`, so there are no breaking changes to any previous stable release!

**For people running a 0.x prerelease**, these are the breaking changes between the latest pre-release and this stable release:

- For boolean settings like `CAMO_ALLOW_IMAGE`, you have to explicitly set `true` or `false` as values, and other values will be rejected as invalid. The previous behavior, where any specified value (like `yes` or confusingly even `no`) would be parsed as `true` has been removed. Omitting those fields will, however, still set them to `false` by default.

# 0.3.0

- You can now set the log output level with `--log-level`/`CAMO_LOG_LEVEL`, and the log output format with `--log-format`/`CAMO_LOG_FORMAT`. Please see [the documentation](/docs/configuration.md) for full details!

# 0.2.0

- `camo-rs` now refuses to start if no content-types are allowed. Before that, Camo would start up just fine, but reject everything, which can be confusing.
- When receiving a status code outside the expected range (`[200..399]`), Camo will still reject that request, but will pass the upstream status code to the client.
- Even with `RUST_LOG=warn`, the log will still contain the encoded digest, target, and - if available - the decoded and validated target URL. This makes debugging production setups easier.

# 0.1.0

This is the first public release. While it is not considered "production-ready" yet, it should work, and is currently being tested in experimental rollouts.


src/settings.rs
//! The Application Settings Module(tm)

use tracing::Level;

/// Specifies the log's output format
#[derive(clap::ValueEnum, Clone, Debug)]
pub enum LogFormat {
    /// Logs in a human-readable format.
    Text,

    /// Logs in a human-readable format, but with colors.
    TextColor,

    /// Logs into a machine-readable JSONL format.
    Json,
}

/// Specifies how much log output Camo generates
#[derive(clap::ValueEnum, Clone, Debug)]
pub enum LogLevel {
    /// Does not log anything at all
    Quiet,

    /// Logs upstream errors and blocked upstream requests
    Warn,

    /// Everything that `warn` includes, but also logs invalid URLs/HMACs
    Info,
}

impl LogLevel {
    pub fn tracing_level(&self) -> Level {
        match self {
            LogLevel::Quiet => Level::ERROR,
            LogLevel::Warn => Level::WARN,
            LogLevel::Info => Level::INFO,
        }
    }
}

/// Application Settings Struct, designed to be primarily used by Clap
#[derive(clap::Parser, Clone, Debug)]
#[clap(about, author, version = env!("CAMO_RS_VERSION"))]
pub struct Settings {
    /// If present, `audio/*` MIME types will be allowed
    #[clap(long = "allow-audio", env = "CAMO_ALLOW_AUDIO")]
    pub allow_audio: bool,

    /// If present, `image/*` MIME types will be allowed
    #[clap(long = "allow-image", env = "CAMO_ALLOW_IMAGE")]
    pub allow_image: bool,

    /// If present, `video/*` MIME types will be allowed
    #[clap(long = "allow-video", env = "CAMO_ALLOW_VIDEO")]
    pub allow_video: bool,

    /// If present, all (including empty) MIME types willb e allowed
    #[clap(long = "allow-all-types", env = "CAMO_ALLOW_ALL_TYPES")]
    pub allow_all_types: bool,

    /// The string used to identify this instance in upstream requests in Via and User-Agent
    #[clap(
        long = "header-via",
        env = "CAMO_HEADER_VIA",
        default_value = "camo-rs asset proxy (+https://github.com/denschub/camo-rs)"
    )]
    pub header_via: String,

    /// Randomly generated string used as a key for calculating the HMAC digest
    #[clap(long = "key", env = "CAMO_KEY")]
    pub key: String,

    /// The maximum `content-length`
    #[clap(
        long = "length-limit",
        env = "CAMO_LENGTH_LIMIT",
        default_value_t = 52428800
    )]
    pub length_limit: usize,

    /// IP and Port this instance should listen on
    #[clap(long = "listen", env = "CAMO_LISTEN", default_value = "[::]:8081")]
    pub listen: String,

    /// Defines how the log output will be formatted
    #[clap(value_enum, long = "log-format", env = "CAMO_LOG_FORMAT", default_value_t = LogFormat::TextColor)]
    pub log_format: LogFormat,

    /// The amount of log output. See the documentation for details
    #[clap(value_enum, long = "log-level", env = "CAMO_LOG_LEVEL", default_value_t = LogLevel::Quiet)]
    pub log_level: LogLevel,

    /// URL, including a trailing slash, relative to the domain Camo is running
    /// on
    ///
    /// For example, if Camo is available on `example.com/camo/`, set this
    /// to `/camo/`. For installations that do not run in a subdirectory, set
    /// this to `/`.
    #[clap(long = "root-url", env = "CAMO_ROOT_URL")]
    pub root_url: String,

    /// Limits the number of threads used - defaults to the number of CPU cores
    #[clap(long, env = "CAMO_THREADS")]
    pub threads: Option<usize>,

    /// The number of seconds to wait for an upstream response
    #[clap(
        long = "upstream-timeout",
        env = "CAMO_UPSTREAM_TIMEOUT",
        default_value_t = 10
    )]
    pub upstream_timeout: usize,
}
